openprescribing/dmd2/search.py

```diff
@@ -46,3 +52,254 @@ def search_by_snomed_code(q):
         return [{"cls": cls, "objs": [obj]}]
 
     return []
+
+
+def advanced_search(cls, search_params):
+    search = json.loads(search_params["search"])
+    include = search_params["include"]
+
+    rules = _build_rules(search)
+    query_obj = _build_query_obj(cls, search)
+
+    qs = cls.objects
+    if "invalid" not in include:
+        qs = qs.valid()
+    if "unavailable" not in include:
+        qs = qs.available()
+    if "no_bnf_code" not in include:
+        qs = qs.with_bnf_code()
+
+    # 10,000 is an arbitrary cut off.  We should do pagination properly.
+    objs = qs.filter(query_obj)[:10001]
+
+    if len(objs) == 10001:
+        too_many_results = True
+        objs = objs[:10000]
+    else:
+        too_many_results = False
+
+    return {"objs": objs, "rules": rules, "too_many_results": too_many_results}
+
+
+def _build_rules(search):
+    """Return structure used to populate a QueryBuilder instance.
+
+    See https://querybuilder.js.org/#method-setRules
+    """
+
+    rules = _build_rules_helper(search)
+    if "condition" not in rules:
+        rules = {"condition": "AND", "rules": [rules]}
+    return rules
+
+
+def _build_rules_helper(search):
+    assert len(search) in [2, 3]
+
+    if len(search) == 2:
+        return {
+            "condition": search[0].upper(),
+            "rules": map(_build_rules_helper, search[1]),
+        }
+    else:
+        value = search[2]
+        if value is True:
+            value = 1
+        elif value is False:
+            value = 0
+
+        return {"id": search[0], "operator": search[1], "value": value}
+
+
+def _build_query_obj(cls, search):
+    """Return Q object to filter dm+d objects based on search.
+    """
+
+    def _build_query_obj_helper(search):
+        assert len(search) in [2, 3]
+
+        if len(search) == 2:
+            fn = {"and": Q.__and__, "or": Q.__or__}[search[0]]
+            clauses = map(_build_query_obj_helper, search[1])
+            return reduce(fn, clauses, Q())
+        else:
+            field_name, operator, value = search
+            if field_name == "bnf_code":
+                if operator == "begins_with":
+                    return Q(bnf_code__startswith=value)
+                elif operator == "not_begins_with":
+                    return ~Q(bnf_code__startswith=value)
+                else:
+                    assert False, operator
+            else:
+                key = _build_lookup_key(cls, field_name, operator)
+                kwargs = {key: value}
+                return Q(**kwargs)
+
+    return _build_query_obj_helper(search)
+
+
+def _build_lookup_key(cls, field_name, operator):
```

Comment:
I wonder if moving the lookup and search filter helpers to a separate module might be useful for readability, so that the entry points are more obvious?  Quite a lot of functions to scan through here. Just a thought.